### DISCUSSION TITLE
Adding new device PJ-1203 (almost there but need a little help)

### DISCUSSION THREAD

**Discussion by curtash** [Koenkk/zigbee2mqtt#22933] [Devices]
Hi, I have followed the [instructions](https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html) but I'm not getting energy readings and I'm unsure why.

The PJ-1203 is an energy meter device with 1 clamp there is a similar one that works PJ-1203**A** I was hoping to do a bit of copy and paste but they are quite different.

I've found the tuyaDatapoints and added the .js file which is being read at least in part because I can see details in the UI.

Data points found using the tuya platform and script:

```
{"18":"Current",
"19":"Power",
"20":"Voltage",
"22":"voltage coe",
"23":"electric coe",
"24":"power coe",
"25":"electricity coe",
"101":"KWH"}
```

![Screenshot 2024-06-04 at 13 22 13](https://github.com/Koenkk/zigbee2mqtt/assets/16105978/638df1a7-027e-4aa3-9919-5a6793965b35)
![Screenshot 2024-06-04 at 13 22 55](https://github.com/Koenkk/zigbee2mqtt/assets/16105978/5f7bc4fd-fda0-45a2-9e1e-c76dd4563578)

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = 
{
    fingerprint: tuya.fingerprint('TS0601', ['_TZE284_cjbofhxw']),
    model: 'PJ-1203',
    vendor: 'TuYa',
    description: 'Energy meter with 80A current clamp',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: 
    [
        // Here you should put all functionality that your device exposes
        e.voltage(), 
        e.power(), 
        e.current(),
        e.energy(),
    ],
    meta: 
    {
        // All datapoints go in here
        tuyaDatapoints: 
        [
            [20, 'voltage', tuya.valueConverter.raw],
            [19, 'power', tuya.valueConverter.raw],
            [18, 'current', tuya.valueConverter.raw],
            [101, 'energy', tuya.valueConverter.raw],
        ],
    },
};
module.exports = definition;
```
Any help would be much appreciated as I'm not a developer so this is all over my head but I'm happy to learn. Hopefully I can sort this as I have several other devices I would like to add. Many thanks.

**Comment by Koenkk** [comment]
`_TZE284_cjbofhxw` is already supported in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by curtash** [reply]
Doh, sorry about that, testing a product in Edge is usually the first thing I try, I thought I had but obviously not! I could have saved myself some time although I have learnt a lot while trying which is good.

I searched here for currently supported items https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/tuya.ts

Where are the dev branch files located? I would like to know so I can just search for the device in the dev branch to see if it is supported "_TZE284_cjbofhxw" without installing Edge. Also I can check what the developer did right and I wrong by comparing files.

On a final note I have noticed that there is not device page associated with _TZE284_cjbofhxw. I assume this needs to be added before it can be included in the stable build? I would be happy to give it a go if someone is not already trying. 

Many thanks for your time and your contribution to z2m!

**Reply by curtash** [reply]
ahh I found it in the place I was looking originally (link above). Originally I searched for PJ-1203 in the file which I now realise was wrong I should have searched **_TZE284_cjbofhxw** 

My only question remains now is what is the process for getting this device into the stable build? Sorry for all the questions I'm just trying to understand how it works, I've done a lot of reading and testing but I'm still a bit lost.

Below is the code that I found in the link that works which I assume the Edge version is using. It looks like mine is almost the same so I'm not sure why it did not work at all.

```
 {
        fingerprint: tuya.fingerprint('TS0601', ['_TZE204_cjbofhxw', '_TZE284_cjbofhxw']),
        model: 'PJ-MGW1203',
        vendor: 'Tuya',
        description: 'Clamp meter',
        fromZigbee: [tuya.fz.datapoints, tuya.fz.gateway_connection_status],
        toZigbee: [tuya.tz.datapoints],
        configure: tuya.configureMagicPacket,
        exposes: [e.current(), e.power(), e.voltage(), e.energy()],
        meta: {
            tuyaDatapoints: [
                [18, 'current', tuya.valueConverter.divideBy1000],
                [19, 'power', tuya.valueConverter.divideBy10],
                [20, 'voltage', tuya.valueConverter.divideBy10],
                [101, 'energy', tuya.valueConverter.divideBy1000],
            ],
        },
        whiteLabel: [
            tuya.whitelabel('Tuya', 'PJ-1203-W', 'Electricity energy monitor', ['_TZE284_cjbofhxw']),
        ],
    },
```

**Reply by Koenkk** [reply]
> My only question remains now is what is the process for getting this device into the stable build? 

It will be included in the next release, which is every 1st of the month.
